File: lexodd_app/lib/models/employee.dart
class Employee {
  final String? id;
  final String? employeeId;
  final String firstName;
  final String lastName;
  final String email;
  final String phone;
  final String? alternatePhone;
  final DateTime? dateOfBirth;
  final String gender;
  final String? bloodGroup;
  final String? maritalStatus;
  final String? profilePhoto;
  final Address? address;
  final Address? permanentAddress;
  final bool sameAsPermanent;
  final String department;
  final String designation;
  final DateTime? dateOfJoining;
  final String employmentType;
  final String? workLocation;
  final String? reportingManager;
  final double? ctc;
  final EmergencyContact? emergencyContact;
  final BankDetails? bankDetails;
  final DocumentDetails? documents;
  final List<Education>? education;
  final LeaveBalance? leaveBalance;
  final String? role;
  final bool isEmailVerified;
  final bool isActive;
  final String approvalStatus;
  final DateTime? approvalDate;
  final String? rejectionReason;

  String get fullName => '$firstName $lastName';
  String get initials =>
      '${firstName.isNotEmpty ? firstName[0] : ''}${lastName.isNotEmpty ? lastName[0] : ''}'
          .toUpperCase();

  int? get age {
    if (dateOfBirth == null) return null;
    final today = DateTime.now();
    int age = today.year - dateOfBirth!.year;
    if (today.month < dateOfBirth!.month ||
        (today.month == dateOfBirth!.month && today.day < dateOfBirth!.day)) {
      age--;
    }
    return age;
  }

  String get tenure {
    if (dateOfJoining == null) return 'N/A';
    final today = DateTime.now();
    int totalMonths = (today.year - dateOfJoining!.year) * 12 +
        (today.month - dateOfJoining!.month);
    int years = totalMonths ~/ 12;
    int months = totalMonths % 12;
    return years > 0 ? '$years yr $months mo' : '$months months';
  }

  Employee({
    this.id,
    this.employeeId,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phone,
    this.alternatePhone,
    this.dateOfBirth,
    required this.gender,
    this.bloodGroup,
    this.maritalStatus,
    this.profilePhoto,
    this.address,
    this.permanentAddress,
    this.sameAsPermanent = false,
    required this.department,
    required this.designation,
    this.dateOfJoining,
    required this.employmentType,
    this.workLocation,
    this.reportingManager,
    this.ctc,
    this.emergencyContact,
    this.bankDetails,
    this.documents,
    this.education,
    this.leaveBalance,
    this.role,
    this.isEmailVerified = false,
    this.isActive = true,
    this.approvalStatus = 'pending',
    this.approvalDate,
    this.rejectionReason,
  });

  factory Employee.fromJson(Map<String, dynamic> json) {
    return Employee(
      id: json['_id'] ?? json['id'],
      employeeId: json['employeeId'],
      firstName: json['firstName'] ?? '',
      lastName: json['lastName'] ?? '',
      email: json['email'] ?? '',
      phone: json['phone'] ?? '',
      alternatePhone: json['alternatePhone'],
      dateOfBirth: json['dateOfBirth'] != null
          ? DateTime.tryParse(json['dateOfBirth'].toString())
          : null,
      gender: json['gender'] ?? '',
      bloodGroup: json['bloodGroup'],
      maritalStatus: json['maritalStatus'],
      profilePhoto: json['profilePhoto'],
      address:
          json['address'] != null ? Address.fromJson(json['address']) : null,
      permanentAddress: json['permanentAddress'] != null
          ? Address.fromJson(json['permanentAddress'])
          : null,
      sameAsPermanent: json['sameAsPermanent'] ?? false,
      department: json['department'] ?? '',
      designation: json['designation'] ?? '',
      dateOfJoining: json['dateOfJoining'] != null
          ? DateTime.tryParse(json['dateOfJoining'].toString())
          : null,
      employmentType: json['employmentType'] ?? '',
      workLocation: json['workLocation'],
      reportingManager: json['reportingManager'],
      ctc: json['ctc'] is num ? (json['ctc'] as num).toDouble() : null,
      emergencyContact: json['emergencyContact'] != null
          ? EmergencyContact.fromJson(json['emergencyContact'])
          : null,
      bankDetails: json['bankDetails'] != null
          ? BankDetails.fromJson(json['bankDetails'])
          : null,
      documents: json['documents'] != null
          ? DocumentDetails.fromJson(json['documents'])
          : null,
      education: json['education'] != null
          ? (json['education'] as List)
              .map((e) => Education.fromJson(e))
              .toList()
          : null,
      leaveBalance: json['leaveBalance'] != null
          ? LeaveBalance.fromJson(json['leaveBalance'])
          : null,
      role: json['role'],
      isEmailVerified: json['isEmailVerified'] ?? false,
      isActive: json['isActive'] ?? true,
      approvalStatus: json['approvalStatus'] ?? 'pending',
      approvalDate: json['approvalDate'] != null
          ? DateTime.tryParse(json['approvalDate'].toString())
          : null,
      rejectionReason: json['rejectionReason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'phone': phone,
      'alternatePhone': alternatePhone,
      'dateOfBirth': dateOfBirth?.toIso8601String(),
      'gender': gender,
      'bloodGroup': bloodGroup,
      'profilePhoto': profilePhoto,
      'maritalStatus': maritalStatus,
      'address': address?.toJson(),
      'permanentAddress': permanentAddress?.toJson(),
      'sameAsPermanent': sameAsPermanent,
      'department': department,
      'designation': designation,
      'dateOfJoining': dateOfJoining?.toIso8601String(),
      'employmentType': employmentType,
      'workLocation': workLocation,
      'reportingManager': reportingManager,
      'ctc': ctc,
      'emergencyContact': emergencyContact?.toJson(),
      'bankDetails': bankDetails?.toJson(),
      'documents': documents?.toJson(),
      'education': education?.map((e) => e.toJson()).toList(),
      'role': role,
      'isEmailVerified': isEmailVerified,
      'isActive': isActive,
      'approvalStatus': approvalStatus,
      'approvalDate': approvalDate?.toIso8601String(),
      'rejectionReason': rejectionReason,
    };
  }
}

class Address {
  final String? street, city, state, pincode, country;
  Address({this.street, this.city, this.state, this.pincode, this.country});
  factory Address.fromJson(Map<String, dynamic> json) => Address(
        street: json['street'],
        city: json['city'],
        state: json['state'],
        pincode: json['pincode'],
        country: json['country'],
      );
  Map<String, dynamic> toJson() => {
        'street': street,
        'city': city,
        'state': state,
        'pincode': pincode,
        'country': country,
      };
  String get fullAddress => [street, city, state, pincode, country]
      .where((e) => e != null && e.isNotEmpty)
      .join(', ');
}

class EmergencyContact {
  final String? name, relationship, phone;
  EmergencyContact({this.name, this.relationship, this.phone});
  factory EmergencyContact.fromJson(Map<String, dynamic> json) =>
      EmergencyContact(
        name: json['name'],
        relationship: json['relationship'],
        phone: json['phone'],
      );
  Map<String, dynamic> toJson() =>
      {'name': name, 'relationship': relationship, 'phone': phone};
}

class BankDetails {
  final String? accountNumber, bankName, branchName, ifscCode, accountType;
  BankDetails(
      {this.accountNumber,
      this.bankName,
      this.branchName,
      this.ifscCode,
      this.accountType});
  factory BankDetails.fromJson(Map<String, dynamic> json) => BankDetails(
        accountNumber: json['accountNumber'],
        bankName: json['bankName'],
        branchName: json['branchName'],
        ifscCode: json['ifscCode'],
        accountType: json['accountType'],
      );
  Map<String, dynamic> toJson() => {
        'accountNumber': accountNumber,
        'bankName': bankName,
        'branchName': branchName,
        'ifscCode': ifscCode,
        'accountType': accountType,
      };
}

class DocumentDetails {
  final String? aadharNumber, panNumber, passportNumber, drivingLicense;
  DocumentDetails(
      {this.aadharNumber,
      this.panNumber,
      this.passportNumber,
      this.drivingLicense});
  factory DocumentDetails.fromJson(Map<String, dynamic> json) =>
      DocumentDetails(
        aadharNumber: json['aadharNumber'],
        panNumber: json['panNumber'],
        passportNumber: json['passportNumber'],
        drivingLicense: json['drivingLicense'],
      );
  Map<String, dynamic> toJson() => {
        'aadharNumber': aadharNumber,
        'panNumber': panNumber,
        'passportNumber': passportNumber,
        'drivingLicense': drivingLicense,
      };
}

class Education {
  final String? degree, institution, university;
  final int? yearOfPassing;
  final double? percentage;
  Education(
      {this.degree,
      this.institution,
      this.university,
      this.yearOfPassing,
      this.percentage});
  factory Education.fromJson(Map<String, dynamic> json) => Education(
        degree: json['degree'],
        institution: json['institution'],
        university: json['university'],
        yearOfPassing: json['yearOfPassing'],
        percentage: json['percentage']?.toDouble(),
      );
  Map<String, dynamic> toJson() => {
        'degree': degree,
        'institution': institution,
        'university': university,
        'yearOfPassing': yearOfPassing,
        'percentage': percentage,
      };
}

class LeaveBalance {
  final double casual, sick, earned, maternity, paternity, compOff;
  LeaveBalance(
      {this.casual = 12,
      this.sick = 12,
      this.earned = 15,
      this.maternity = 0,
      this.paternity = 0,
      this.compOff = 0});
  factory LeaveBalance.fromJson(Map<String, dynamic> json) => LeaveBalance(
        casual: (json['casual'] ?? 12).toDouble(),
        sick: (json['sick'] ?? 12).toDouble(),
        earned: (json['earned'] ?? 15).toDouble(),
        maternity: (json['maternity'] ?? 0).toDouble(),
        paternity: (json['paternity'] ?? 0).toDouble(),
        compOff: (json['compOff'] ?? 0).toDouble(),
      );
  double get total => casual + sick + earned + compOff;
}

File: hrms_backend/models/Employee.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');

const employeeSchema = new mongoose.Schema({
  // Personal Info
  employeeId: { type: String, unique: true, sparse: true },
  firstName: { type: String, required: [true, 'First name required'], trim: true, maxlength: 50 },
  lastName: { type: String, required: [true, 'Last name required'], trim: true, maxlength: 50 },
  email: { 
    type: String, required: [true, 'Email required'], unique: true, lowercase: true, trim: true,
    match: [/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/, 'Invalid email']
  },
  phone: { type: String, required: [true, 'Phone required'], match: [/^[6-9]\d{9}$/, 'Invalid phone'] },
  alternatePhone: { type: String },
  dateOfBirth: { type: Date, required: [true, 'DOB required'] },
  gender: { type: String, required: true, enum: ['Male', 'Female', 'Other', 'Prefer not to say'] },
  bloodGroup: { type: String, enum: ['A+', 'A-', 'B+', 'B-', 'AB+', 'AB-', 'O+', 'O-', ''] },
  maritalStatus: { type: String, enum: ['Single', 'Married', 'Divorced', 'Widowed', ''] },
  profilePhoto: { type: String, default: '' },

  // Address
  address: {
    street: String, city: String, state: String,
    pincode: { type: String, match: [/^\d{6}$/, 'Invalid pincode'] },
    country: { type: String, default: 'India' }
  },
  permanentAddress: {
    street: String, city: String, state: String,
    pincode: String,
    country: { type: String, default: 'India' }
  },
  sameAsPermanent: { type: Boolean, default: false },

  // Employment
  department: {
    type: String, required: [true, 'Department required'],
    enum: ['Engineering', 'Human Resources', 'Finance', 'Marketing', 'Sales', 'Operations', 'Customer Support', 'Design', 'Product', 'Legal', 'Administration', 'IT', 'Research', 'Other']
  },
  designation: { type: String, required: [true, 'Designation required'], trim: true },
  dateOfJoining: { type: Date, required: [true, 'Joining date required'] },
  employmentType: { type: String, required: true, enum: ['Full-time', 'Part-time', 'Contract', 'Intern', 'Freelance'] },
  workLocation: { type: String, enum: ['Office', 'Remote', 'Hybrid'], default: 'Office' },
  reportingManager: { type: String, trim: true },
  ctc: { type: Number, min: 0 },

  // Emergency Contact
  emergencyContact: {
    name: String,
    relationship: String,
    phone: String
  },

  // Bank Details
  bankDetails: {
    accountNumber: String,
    bankName: String,
    branchName: String,
    ifscCode: { type: String, uppercase: true },
    accountType: { type: String, enum: ['Savings', 'Current', ''] }
  },

  // Documents
  documents: {
    aadharNumber: String,
    panNumber: { type: String, uppercase: true },
    passportNumber: String,
    drivingLicense: String
  },

  // Education
  education: [{
    degree: String,
    institution: String,
    university: String,
    yearOfPassing: Number,
    percentage: Number
  }],

  // Leave Balance
  leaveBalance: {
    casual: { type: Number, default: 12 },
    sick: { type: Number, default: 12 },
    earned: { type: Number, default: 15 },
    maternity: { type: Number, default: 0 },
    paternity: { type: Number, default: 0 },
    compOff: { type: Number, default: 0 }
  },

  // Auth
  password: { type: String, required: true, minlength: 8, select: false },
  role: { type: String, enum: ['employee', 'manager', 'hr', 'admin'], default: 'employee' },
  isEmailVerified: { type: Boolean, default: false },
  isActive: { type: Boolean, default: true },
  passwordChangedAt: Date,
  lastLogin: Date,
  loginAttempts: { type: Number, default: 0 },
  lockUntil: Date

}, { timestamps: true, toJSON: { virtuals: true }, toObject: { virtuals: true } });

// Virtuals
employeeSchema.virtual('fullName').get(function() {
  return `${this.firstName} ${this.lastName}`;
});

employeeSchema.virtual('age').get(function() {
  if (!this.dateOfBirth) return null;
  const today = new Date();
  const birth = new Date(this.dateOfBirth);
  let age = today.getFullYear() - birth.getFullYear();
  if (today.getMonth() < birth.getMonth() || (today.getMonth() === birth.getMonth() && today.getDate() < birth.getDate())) age--;
  return age;
});

employeeSchema.virtual('tenure').get(function() {
  if (!this.dateOfJoining) return 'N/A';
  const months = (new Date().getFullYear() - this.dateOfJoining.getFullYear()) * 12 + (new Date().getMonth() - this.dateOfJoining.getMonth());
  const years = Math.floor(months / 12);
  const rem = months % 12;
  return years > 0 ? `${years} yr ${rem} mo` : `${rem} months`;
});

employeeSchema.virtual('isLocked').get(function() {
  return !!(this.lockUntil && this.lockUntil > Date.now());
});

// Hash password before save
employeeSchema.pre('save', async function(next) {
  if (!this.isModified('password')) return next();
  this.password = await bcrypt.hash(this.password, 12);
  next();
});

// Generate employee ID
employeeSchema.pre('save', async function(next) {
  if (this.employeeId) return next();
  const count = await mongoose.model('Employee').countDocuments();
  const year = new Date().getFullYear().toString().slice(-2);
  const month = (new Date().getMonth() + 1).toString().padStart(2, '0');
  this.employeeId = `EMP${year}${month}${(count + 1).toString().padStart(4, '0')}`;
  next();
});

// Update passwordChangedAt
employeeSchema.pre('save', function(next) {
  if (!this.isModified('password') || this.isNew) return next();
  this.passwordChangedAt = Date.now() - 1000;
  next();
});

// Methods
employeeSchema.methods.comparePassword = async function(candidatePassword) {
  return await bcrypt.compare(candidatePassword, this.password);
};

employeeSchema.methods.getSignedJwtToken = function() {
  return jwt.sign(
    { id: this._id, role: this.role, email: this.email },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRE }
  );
};

employeeSchema.methods.incrementLoginAttempts = async function() {
  if (this.lockUntil && this.lockUntil < Date.now()) {
    return await this.updateOne({ $set: { loginAttempts: 1 }, $unset: { lockUntil: 1 } });
  }
  const updates = { $inc: { loginAttempts: 1 } };
  if (this.loginAttempts + 1 >= 5 && !this.isLocked) {
    updates.$set = { lockUntil: Date.now() + 30 * 60 * 1000 };
  }
  return await this.updateOne(updates);
};

employeeSchema.methods.resetLoginAttempts = async function() {
  return await this.updateOne({
    $set: { loginAttempts: 0, lastLogin: new Date() },
    $unset: { lockUntil: 1 }
  });
};

module.exports = mongoose.model('Employee', employeeSchema);
